kb/2025-11-04/k8s_vs_longhorn_volumeattachment.md

```diff
@@ -0,0 +1,265 @@
+└─────────────────────┬───────────────────────────────────────┘
+                      │
+                      ▼
+┌─────────────────────────────────────────────────────────────┐
+│ Longhorn CSI Plugin                                         │
+│ Returns attach success to external-attacher                 │
+└─────────────────────┬───────────────────────────────────────┘
+                      │
+                      ▼
+┌─────────────────────────────────────────────────────────────┐
+│ CSI External-Attacher                                       │
+│ Updates K8s VolumeAttachment.Status.Attached = true         │
+└─────────────────────────────────────────────────────────────┘
+```
+
+The resulting Longhorn VolumeAttachment YAML:
+
+```yaml
+apiVersion: longhorn.io/v1beta2
+kind: VolumeAttachment
+metadata:
+  name: pvc-0b9c8d59-0ae8-413c-8bc5-af32b932b8ab
+  namespace: longhorn-system
+  labels:
+    longhornvolume: pvc-0b9c8d59-0ae8-413c-8bc5-af32b932b8ab
+spec:
+  volume: pvc-0b9c8d59-0ae8-413c-8bc5-af32b932b8ab
+  attachmentTickets:
+    # This CSI ticket was triggered by K8s VolumeAttachment (Pod binding)
+    csi-3d3120f43480db87c91a6902d670c35899917c03f9f6f81db7bf26d9d66e45ec:
+      id: csi-3d3120f43480db87c91a6902d670c35899917c03f9f6f81db7bf26d9d66e45ec
+      type: csi-attacher
+      nodeID: harvester-node-1
+      parameters:
+        disableFrontend: "false"
+status:
+  attachmentTicketStatuses:
+    csi-3d3120f43480db87c91a6902d670c35899917c03f9f6f81db7bf26d9d66e45ec:
+      id: csi-3d3120f43480db87c91a6902d670c35899917c03f9f6f81db7bf26d9d66e45ec
+      satisfied: true  # Volume successfully attached
+      conditions:
+      - type: Satisfied
+        status: "True"
+```
+
+Notice the `csi-attacher` ticket type - this confirms the attachment was triggered by Kubernetes VolumeAttachment through the CSI flow, not by Longhorn internal operations.
+
+### Trigger Points
+
+Understanding when each VolumeAttachment is created or modified is crucial for troubleshooting attachment issues:
+
+1. **K8s VolumeAttachment Creation**: Triggered when Pod is scheduled to a node requiring a PVC
+   - Managed by Kubernetes Attach/Detach (AD) Controller
+   - One VolumeAttachment per PV-node combination
+   - Represents Kubernetes' intent to attach the volume
+
+2. **Longhorn VolumeAttachment Ticket Addition**: Triggered by various Longhorn components based on operation needs:
+   - **CSI Plugin** - when CSI ControllerPublishVolume is called
+   - **Snapshot Controller** - when creating snapshots of detached volumes
+   - **Backup Controller** - when backing up volumes
+   - **UI/API** - when users manually attach volumes via Longhorn UI
+   - **Clone Controller** - when managing source volume during clone
+   - **Restore Controller** - when restoring data from backups
+   - **Expansion Controller** - when expanding volume size
+   - And other Longhorn controllers based on their specific operational needs
```

Comment:
note: for tickets that are not CSI or UI/API, the corresponding operations should still function while the volume is attached, even if the ticket isn't marked as satisfied. If the volume is detached, those tickets will temporarily attach the volume by priority, execute the operation, and then detach the volume after the operation is finished.

cc @PhanLe1010